docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+    Vector128<int> sum = Vector128<int>.Zero;
+
+    // We need to obtain the reference to first value in the buffer, it's used later for loading vectors from memory.
+    ref int searchSpace = ref MemoryMarshal.GetReference(buffer);
+    // And an offset, that is going to be used by vectorized and scalar loops.
+    nuint elementOffset = 0;
+    // And the last valid offset from which we can load the values
+    nuint oneVectorAwayFromEnd = (nuint)(buffer.Length - Vector128<int>.Count);
+    for (; elementOffset <= oneVectorAwayFromEnd; elementOffset += (nuint)Vector128<int>.Count)
+    {
+        // We load a vector from given offset.
+        Vector128<int> loaded = Vector128.LoadUnsafe(ref searchSpace, elementOffset);
+        // We add 4 integers at a time:
+        sum += loaded;
+    }
+
+    // We sum all 4 integers from the vector to one
+    int result = Vector128.Sum(sum);
+
+    // And handle the remaining elements, in a non-vectorized way:
+    while (elementOffset < (nuint)buffer.Length)
+    {
+        result += buffer[(int)elementOffset];
+        elementOffset++;
+    }
+
+    return result;
+}
+```
+
+**Note:** Use `ref MemoryMarshal.GetReference(span)` instead of `ref span[0]` and `ref MemoryMarshal.GetArrayDataReference(array)` instead of `ref array[0]` to handle empty buffer scenarios (which would throw `IndexOutOfRangeException`). If the buffer is empty, these methods return a reference to the location where the 0th element would have been stored. Such a reference may or may not be null. You can use it for pinning but you must never dereference it.
+
+**Note:** The `GetReference` method has an overload that accepts a `ReadOnlySpan` and returns mutable reference. Please use it with caution! To get a `readonly` reference, you need to use [ReadOnlySpan<T>.GetPinnableReference](https://learn.microsoft.com/dotnet/api/system.readonlyspan-1.getpinnablereference).
+
+**Note:** Please keep in mind that `Vector128.Sum` is a static method. `Vectior128<T>` and `Vector256<T>` provide both instance and static methods (operators like `+` are just static methods in C#). `Vector128` and `Vector256` are non-generic static classes with static methods only. It's important to know about their existence when searching for methods.
+
+### Vectorized remainder handling
+
+Now imagine that we need to check whether the given buffer contains a specific number. In this case, processing some values more than once is acceptable, we don't need to handle the remainder in a non-vectorized fashion.
+
+Example: a buffer contains six 32-bit integers, `Vector128` is accelerated, and it can work with four integers at a time. In the first loop iteration, we handle the first four elements. In the second (and last) iteration we need to handle the remaining two elements. Since the remainder is smaller than one `Vector128` and we are not mutating the input, we perform a vectorized operation on a `Vector128` containing the last four elements.
```

Comment:
Consider:
```suggestion
Example: a buffer contains six 32-bit integers, `Vector128` is accelerated, and it can work with four integers at a time. In the first loop iteration, we handle the first four elements. In the second (and last) iteration we need to handle the remaining two elements. Since the remainder is smaller than one `Vector128` and we are not mutating the input or using the inputs in a way that changes the result, we perform a vectorized operation on a `Vector128` containing the last four elements.
```

The "technical" term is the operation is "idempotent"